sdk/src/vote/submission.rs

```diff
@@ -0,0 +1,587 @@
+        if let Some(unique_key) = &self.most_voted_key {
+            Ok(self
+                .get_submission_weight(store, unique_key)?
+                .unwrap_or_else(TokenAmount::zero))
+        } else {
+            Ok(TokenAmount::zero())
+        }
+    }
+
+    pub fn get_submission<BS: Blockstore>(
+        &self,
+        store: &BS,
+        unique_key: &UniqueBytesKey,
+    ) -> anyhow::Result<Option<Vote>> {
+        let hamt = self.submissions.load(store)?;
+        let key = BytesKey::from(unique_key.as_slice());
+        Ok(hamt.get(&key)?.cloned())
+    }
+
+    pub fn derive_execution_status(
+        &self,
+        total_weight: TokenAmount,
+        most_voted_weight: TokenAmount,
+    ) -> VoteExecutionStatus {
+        let threshold = total_weight
+            .clone()
+            .mul(*RATIO_NUMERATOR)
+            .div_floor(*RATIO_DENOMINATOR);
+
+        // note that we require THRESHOLD to be surpassed, equality is not enough!
+        if self.total_submission_weight <= threshold {
+            return VoteExecutionStatus::ThresholdNotReached;
+        }
+
+        // now we have reached the threshold
+
+        // consensus reached
+        if most_voted_weight > threshold {
+            return VoteExecutionStatus::ConsensusReached;
+        }
+
+        // now the total submissions has reached the threshold, but the most submitted vote
+        // has yet to reach the threshold, that means consensus has not reached.
+
+        // we do a early termination check, to see if consensus will ever be reached.
+        //
+        // consider an example that consensus will never be reached:
+        //
+        // -------- | -------------------------|--------------- | ------------- |
+        //     MOST_VOTED                 THRESHOLD     TOTAL_SUBMISSIONS  TOTAL_WEIGHT
+        //
+        // we see MOST_VOTED is smaller than THRESHOLD, TOTAL_SUBMISSIONS and TOTAL_WEIGHT, if
+        // the potential extra votes any vote can obtain, i.e. TOTAL_WEIGHT - TOTAL_SUBMISSIONS,
+        // is smaller than or equal to the potential extra vote the most voted can obtain, i.e.
+        // THRESHOLD - MOST_VOTED, then consensus will never be reached, no point voting, just abort.
+        if threshold - most_voted_weight >= total_weight - &self.total_submission_weight {
+            VoteExecutionStatus::RoundAbort
+        } else {
+            VoteExecutionStatus::ReachingConsensus
+        }
+    }
+
+    /// Checks if the submitter has already submitted the checkpoint.
+    pub fn has_submitted<BS: Blockstore>(
+        &self,
+        store: &BS,
+        submitter: &Address,
+    ) -> anyhow::Result<bool> {
+        let addr_byte_key = BytesKey::from(submitter.to_bytes());
+        let hamt = self.submitters.load(store)?;
+        Ok(hamt.contains_key(&addr_byte_key)?)
+    }
+}
+
+impl<V: Serialize> Serialize for EpochVoteSubmissions<V> {
```

Comment:
Somehow if I use the derive, it would raise a trait not constraint error from Rust which I have already constraint. Only this way I can bypass the compilation error.